adding isSecret method to mask values

aksm-ms opened the pull request:
#16 
Adding code changes to avoid masking input which is not provided as secrets. Code changes will address the issue where if the input json is given as plain text, action does not add it as secret so that the values in settings are not masked by actions platform.

Currently the inputs "app-settings-json" and "connection-strings-json" are masked. Individual key value pairs are masked. 
Input "general-settings-json" is not masked.

If the user gives "app-settings-json" and "connection-strings-json" as plain text and not as secrets, with this change we expect the json to have the following structure for not masking:

```
app-settings-json': `[
        {
            "name": "key2",
            "value": "value",
            "slotSetting": true
        }, 
        {
            "secret": false
        }
    ]

```
If "app-settings-json" and "connection-strings-json" have { "secret": false }, action will not mask the input json. Otherwise we will mask all the values in "app-settings-json" and "connection-strings-json".

@N-Usha Please let me know if "general-settings-json" input should also be masked. Currently action does not mask the general-settings-json or the its individual key-value pairs.

N-Usha commented:
@aksm-ms  I dont see the need to change anything for general-settings-json. Particularly since there was no customer ask around it. Thanks for following through this.

asranja commented:
@aksm-ms  app-settings-json seems to expect an array of objects with properties `name, value, slotSetting`. Adding a new object ```  {
            "secret": false
        }``` in the array does not seem very-intuitive. We are adding objects of two different types to the same array.

- We can have a secret property in the object itself which can determine if the individual key needs to be masked. 

Ex:
```
app-settings-json': [
        {
            "name": "key2",
            "value": "value",
            "slotSetting": true,
            "secret": false
        }
```
- Can we have a new input ? 
- Can we depend on an environment variable. Ex: MaskAppSettings

aksm-ms commented:
@asranja In general, anything which is set as secret is masked by actions platform in subsequent actions in a workflow.
Ideally we should avoid setting any structured data(like json) as secret. Please refer [Encrypted secrets](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets#naming-your-secrets)

I found out through experiments that the reason for this is if you are setting a json as a secret, the complete json will be masked in logs but individual values will not be masked, which is a security risk. In appservice-settings action, we cannot avoid taking input as json because app settings on the portal itself is a json array. 
To avoid the security risk, in action we mask individual values for app-settings-json and connections-string-json.

> * We can have a secret property in the object itself which can determine if the individual key needs to be masked.
>

The idea behind adding { "secret": false } in the array and not in individual objects is:

- If the input json is a secret variable, user need not add { "secret": false }  in their json input. All the individual values will be masked by default in action.

- If input json is not a secret variable but plain text, user can add { "secret": false }  as the last element so that we will not mask the individual values, which will avoid any unexpected redaction of logs (#16)

Since there are only 2 cases, secret variable or not, adding { "secret": false } once was sufficient IMO. 

> * Can we have a new input ?
Action takes 3 settings json inputs out of which app-settings and connection-string-json are masked by default currently.

- if we take a new input, say isSecret:True/False, we need to set both the json as secret/not a secret. To avoid this, we should create 2 inputs and if we add any new settings json input support, we need to add an additional input which is not scalable.

> * Can we depend on an environment variable. Ex: MaskAppSettings
If we take an environment variable in the workflow, how are we going to read it in the action? I do not see a read-env variable method in toolkit [(core)](https://github.com/actions/toolkit/tree/main/packages/core). In that case, if we set an environment variable, it must be taken into action via an input variable.

aksm-ms commented:
@asranja As discussed, changed approach to add a new input "mask-inputs". If mask-inputs is set to "false", app-settings-json and connection-strings-json will not be masked. Default for mask-inputs is "true"
cc: @N-Usha

aksm-ms commented:
@asranja I have also refactored code a bit, moving validateSettings and maskValues into separate Utils class. Added unit tests as well. Please review.

asranja commented on action.yml:
```diff
@@ -17,6 +17,10 @@ inputs:
   general-settings-json: #id of input
     description: 'General configuration settings using dictionary syntax - Key Value pairs'
     required: false
+  mask-inputs: 
```
close the name and description with PMs

aksm-ms replied:
@N-Usha Can you please review this?

N-Usha replied:
name looks ok.. But would this apply to all inputs including general-settings? Else needs to be called out in description.

aksm-ms replied:
@N-Usha Currently in v1 general-settings is not masked. But in master it is masked. Hence in this change with mask-Inputs, it will be masked.
if mask-inputs is set as "false" by user, do not mask any settings
if mask-inputs is not set by user, default is true and all 3 settings will be masked (including general-settings)
if mask-inputs is set to anything other than "false", all 3 inputs will be masked.
Please let me know should i update this in action.yml

N-Usha replied:
got it. Hope the value 'false' specified is not case sensitive.. If not, should be called out in description again.

aksm-ms replied:
@N-Usha It is not case sensitive.

asranja commented on src/Utils.ts:
```diff
@@ -0,0 +1,23 @@
+    static validateSettings(customSettings: string, maskInputs?: string) {
+        try {
+            var customParsedSettings = JSON.parse(customSettings);
+            if (!!maskInputs && maskInputs !== "false") {
```
this should be an || instead of &&

asranja replied:
if input does not exist or if it exists, it should not be false.

asranja replied:
```
if (!maskInputs || maskInputs !== "false")
```

aksm-ms replied:
i had added !!maskInputs to check if its not undefined since i had it as an optional argument.
in master general-settings is not an optional argument but in releases/v1 it is, that is why i added optional argument.
since we need to merge code changes from master into releases/v2, i am removing optional argument and removing the !!maskInputs condition as well

asranja reviewed (approved):
Please address the comments before closing.
